documentation/man/einit-log.8
.\" This manpage has been automatically generated by docbook2man 
.\" from a DocBook document.  This tool can be found at:
.\" <http://shell.ipoline.com/~elmert/comp/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "EINIT-LOG" "8" "24 February 2008" "einit-log-0.40.0" ""

.SH NAME
einit-log \- retrieve log data from einit
.SH SYNOPSIS

\fBeinit-log\fR [ \fB-a \fIaddress\fB\fR ] [ \fB [ -h | --help ]\fR ] [ \fB [ -f | --follow ]\fR ] [ \fB [ -r | --replay-only ]\fR ]

.SH "DESCRIPTION"
.PP
\fBeinit-log\fR is a simple command-line utility that may be used to extra log data from
a running einit core. It simply connects to einit, fetches the event log and displays it.
.SH "OPTIONS"
.TP
\fB-a \fIaddress\fB\fR
The 9P2000 address to connect to. The default is \fBunix!/dev/einit-9p\fR\&.
.TP
\fB-h\fR
.TP
\fB--help\fR
Display help screen and exit.
.TP
\fB-r\fR
.TP
\fB--replay-only\fR
Replay the event log, then exit. This is the default behaviour.
.TP
\fB-f\fR
.TP
\fB--follow\fR
Follow the event log, that is, replay first, then stay connected and wait for updates.
.SH "ENVIRONMENT"
.PP
Some command line options may be used to set defaults for this programme.
.TP
\fBEINIT_9P_ADDRESS=\fIaddress\fB\fR
Equivalent to the -a option. The command-line option overrides this variable.
.SH "USAGE"
.PP
The main use of this programme is to be able to retrieve notice logs and bootup messages
from einit's core. Usually this is done via \fBeinit-log>logfile\fR\&.
.PP
\fBeinit-log --follow\fR is also quite useful, since it allows trying something
and seeing immediate results in the event log. An instance of this command is usually spawned
by einit on tty11.
.SH "FILES"
.TP
\fB/dev/einit-9p\fR
The unix socket that is used (by default) for the 9p2000(.u) IPC.
.SH "BUGS"
.PP
Please send bug reports to the \fBissue tracker on einit.org\fR\&.
.PP
Remember that this is BETA software: avoid using this on mission-critical systems.
.SH "SEE ALSO"
.PP
einit(8), einit-feedback(8), einit-sysvd(8)
.PP
The official project website <URL:http://einit.org/>
.SH "AUTHOR"
.PP
Written by Magnus Deininger <mdeininger@jyujin.de>\&.
.SH "COPYRIGHT"
.PP
(C) 2008 Magnus Deininger, All rights reserved.
.PP
Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:
.PP
.TP 0.2i
\(bu
Redistributions of source code must retain the above copyright notice,
this list of conditions and the following disclaimer.
.TP 0.2i
\(bu
Redistributions in binary form must reproduce the above copyright notice,
this list of conditions and the following disclaimer in the documentation
and/or other materials provided with the distribution.
.TP 0.2i
\(bu
Neither the name of the project nor the names of its contributors may be
used to endorse or promote products derived from this software without
specific prior written permission.
.PP
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


ChangeLog
2008/2/18: Ryan Hope
 ? initramfs support
 * timezone support
 ? urandom
 * bootchart support
 * curses feedback daemon
 * better guile support
 * more scheme modules
 * network-v2
 * module-logic-v4
 * logd
 * ipc-9p

2007/11/26: Ryan Hope
 * scheme modules
 * edev
 * switched to git
 * hotplug handler

2007/04/30: mdeininger
 * enhanced feedback
 * fixed some deadlocks

2007/04/24: mdeininger
 i'm getting quite lazy about updating this file... have a look at the svn
 revision history for closer details ;)
 * implemented shutdown-(before|after)
 * fixed problems with before/after (should work again)

2007/04/15: mdeininger
 * more API changes...

2007/04/14: mdeininger
 * API changes... lots of 'em...

2007/04/12: mdeininger
 * changes to the feedback and feedback auto-add code
 * new feedback module for festival (gonna use a thread like with fbsplash)

2007/04/11: mdeininger
 * a number of minor fixes
 * the new core seems to be doing quite well

2007/04/10: mdeininger
 * new core mostly operational.
 * changed configure script to support modules in C++
 * added template for C++ module (ipc-dbus)

2007/04/09: mdeininger
 * started on a new more-or-less event-based module logics core

2007/04/08: mdeininger
 * made suff cleaner by removing the 'feedback' mode, but also auto-adding
   a feedback module if none is specified

2007/04/05: mdeininger
 * added options="feedback" to daemon/shell modules to tag them as feedback-
   modules (gonna be useful for the fbsplash module)
 * moved re-mounting the root-filesystem to mount/critical
 * the fbsplash module seems to work just fine =)
 * moved the control socket to /dev/einit-control
 * removed module reset/reload/zap hooks in favour of "custom" hooks
 * changed mount/* to mount-

2007/04/03: mdeininger
 * parsed some of the configuration in the wrong function/forgot updates
 * "list services" has been moved to the module-logics core, + it will now
   display candidates in the order that einit will try to deal with them.
 * added envfile="" support
 * added is-up="" and is-down="" attrs to daemons -- they run after the daemon
   is forked -- if the return value is shell-ish true, then the appropriate
   change to the daemon is considered successful, else it is seen as failed
 * added need-files="" to daemons -- just add a list of files that must exist
   or else einit won't even try to bring it up (if the list is available).
 * new module: configuration-network

2007/04/02: mdeininger
 * need to split that configuration up :D
 * use /etc/fstab by default
 * introduced a new readdirfilter() function, we did do that a lot...

2007/04/01: mdeininger
 * started on a custom set of (m|re|c)alloc() functions
 * modified configure script to add files to link into the core

[2006/04/01: mdeininger: release, Version 0.21.0]

[2006/03/27: mdeininger: release, Version 0.20.0]

2007/03/22: mdeininger
 * logging works
 * output is a little clearer

2007/03/21: mdeininger
 * logging kinda works...

2007/03/20: mdeininger
 * stripped einit-control even further

2007/03/19: mdeininger
 * fixed/enhanced the build system
 * added some very simple hashing to the tree-linear code...
   seems to decrease execution time considerably

2007/03/18: mdeininger
 * some good progress with compiling modules into the core

2007/03/14: mdeininger
 * need to be able to compile modules in.

2007/03/13: mdeininger
 * this should get us more "const-correct" :)
 * fixed some annoying compiler warnings
 * moved /etc/einit-control to /etc/einit/control

2007/03/11: mdeininger
 * introduced a new function to check for string matches (strmatch())

2007/03/09: mdeininger
 * added support for ryan's shadow modules

2007/03/07: mdeininger
 * fixes, cleanups, new apach2 module

2007/03/05: mdeininger
 * some new modules, random fixes

2007/03/01: mdeininger
 * the v3 module logics core is now by default external
 * the default expat configuration parser is now external as well.

2007/02/28: mdeininger
 * more preparations to get more stuff out of the core.

2007/02/27: mdeininger
 * shell-gdm should also provide "xdm"
 * the v3 core is now the default
 * fixed some deadlocks in the v3 module-logics core
 * pexec() seems to be less flakey now

2007/02/25: mdeininger
 * some new macros
 * new module logics core is coming along nicely

2007/02/22: mdeininger
 * fixed readfile() problem with .keep files

2007/02/21: mdeininger
 * new new module logics core, again

2007/02/20: mdeininger
 * added service transformations
 * build system changes
 * moved get_plan_progress from the visualiser to the module-logic.c file
 * new sobexsrv module by RedIpS -- nice man ;)

2007/02/19: mdeininger
 * etc/einit/einit.xml has been moved to lib/einit/einit.xml
   (einit.xml should not be modified by users)
 * lib/einit/*.so have been moved to lib/einit/modules
 * added code to read and use the sysctl.conf file
 * ha, found the bloody bugger
 * the xml configuration parser may now create missing include directories
   itself
 * new ipc command "set variable"
 * new ipc command "save configuration"

2007/02/18: mdeininger
 * new function for error reporting (and optional hara-kiri)
 * workover using -Wall
 * added some code to check for errors (that normally need not be checked)

2007/02/16: mdeininger
 * fixed some minor things

2007/02/15: mdeininger
 * extra module for entranceD (additional to the entrance module from earlier)
 * "list services" will now list deprecated modules with a red "-" instead of
   the usual *
 * fixed a stray malloc()

2007/02/08: mdeininger
 * new module for ipc commands that modify the configuration

2007/02/07: mdeininger
 * output for aforementioned ipc call

2007/02/06: mdeininger
 * ipc call to get configuration nodes

[2006/02/05: mdeininger: release, Version 0.16.0]

2007/02/04: mdeininger
 * added a cfg_filter() function

2007/02/02: mdeininger
 * added "rc zap" command
 * einit-control should now auto-add --ansi if appropriate

2007/02/01: mdeininger

src/einit-log.c
/*
 *  einit-log.c
 *  einit
 *
 *  Created by Magnus Deininger on 16/02/2008.
 *  Copyright 2008 Magnus Deininger. All rights reserved.
 *
 */

/*
Copyright (c) 2008, Magnus Deininger
All rights reserved.

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright notice,
          this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright notice,
          this list of conditions and the following disclaimer in the documentation
          and/or other materials provided with the distribution.
    * Neither the name of the project nor the names of its contributors may be
          used to endorse or promote products derived from this software without
          specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/socket.h>
#include <fcntl.h>
#include <string.h>

#include <einit/event.h>

#include <einit/configuration-static.h>

#include <einit/einit.h>

void event_handler_feedback_notice (struct einit_event *ev) {
 if (ev->string)
  fprintf (stdout, "[notice] %i: %s\n", ev->flag, ev->string);
}

void event_handler_update_module_status (struct einit_event *ev) {
 if (ev->string)
  fprintf (stdout, "[%s] %s\n", ev->rid, ev->string);
}

void event_handler_update_service_enabled (struct einit_event *ev) {
 fprintf (stdout, "[%s] enabled\n", ev->string);
}

void event_handler_update_service_disabled (struct einit_event *ev) {
 fprintf (stdout, "[%s] disabled\n", ev->string);
}

void help (char **argv) {
 printf ("Usage: %s [options]\n\n"
         "Options:\n"
         " -f, --follow       Follow the event Log.\n"
         " -n, --replay-only  Stop after all the current events were displayed (default)\n"
         " -h, --help         This Message\n", argv[0]);
}

int main(int argc, char **argv, char **env) {
 char follow = 0;
 int i = 1;

 for (; argv[i]; i++) {
  if (strmatch (argv[i], "-f") || strmatch (argv[i], "--follow")) {
   follow = 1;
  } else if (strmatch (argv[i], "-n") || strmatch (argv[i], "--replay-only")) {
   follow = 0;
  } else if (strmatch (argv[i], "-h") || strmatch (argv[i], "--help")) {
   help(argv);
   return EXIT_SUCCESS;
  } else {
   help(argv);
   return EXIT_FAILURE;
  }
 }

 if (!einit_connect(&argc, argv)) {
  perror ("Could not connect to eINIT");
  sleep (1);
  if (!einit_connect(&argc, argv)) {
   perror ("Could not connect to eINIT, giving up");
   exit (EXIT_FAILURE);
  }
 }

 event_listen (einit_feedback_notice, event_handler_feedback_notice);
 event_listen (einit_feedback_module_status, event_handler_update_module_status);
 event_listen (einit_core_service_enabled, event_handler_update_service_enabled);
 event_listen (einit_core_service_disabled, event_handler_update_service_disabled);

 if (follow)
  einit_event_loop();
 else
  einit_replay_events();

 einit_disconnect();

 return 0;
}
